### ISSUE TITLE
CONNECTIVITY04 - IP prefix diversity requirement 

### ISSUE THREAD

**Issue by d4nys3k** [zonemaster/zonemaster#1174]
Hello,
current implementation of CONNECTIVITY04 has strict requirement not to have two nameservers within single subnet. The idea of diversity, in my opinion, is primarily about having **at least one** nameserver in another network and there is nothing wrong with having two nameservers in single network, as long as I have at least one in another network (_so, you can have two nameservers in single subnet and third in different network, for example_). if it doesn't matter that the zone has only two servers, if each is in a different subnet, 2+1 configuration should not be a problem. 

I tried to modify `lib/Zonemaster/Engine/Test/Connectivity.pm` in this way by patch below to not issue warning in such case. I'm using version 4.7.0 of Engine at this time. It does what I expect based on the thoughts above.

Could you please consider modifying your CONNECTIVITY04 test approach? 

```
--- Connectivity.pm.orig        2023-06-01 10:39:55.000000000 +0200
+++ Connectivity.pm     2023-06-30 23:44:41.247946421 +0200
@@ -615,2 +615,3 @@
     my %prefixes;
+    my %nscount=(4=>0,6=>0);
 
@@ -645,2 +646,3 @@
                 if ( $prefix->version == 4 ) {
+                    $nscount{4}++;
                     $prefix_str = $prefix->prefix;
@@ -648,2 +650,3 @@
                 elsif ( $prefix->version == 6 ) {
+                    $nscount{6}++;
                     $prefix_str = $prefix->short . '/' . $prefix->prefixlen;
@@ -671,6 +674,6 @@
         foreach my $prefix ( keys %{ $prefixes{$ip_version} } ) {
-            if ( scalar @{ $prefixes{$ip_version}{$prefix} } == 1 ) {
+            if ( scalar @{ $prefixes{$ip_version}{$prefix} } <= $nscount{$ip_version}-1 ) {
                 push @combined_ns, @{ $prefixes{$ip_version}{$prefix} };
             }
-            elsif ( scalar @{ $prefixes{$ip_version}{$prefix} } >= 2 ) {
+            elsif ( scalar @{ $prefixes{$ip_version}{$prefix} } == $nscount{$ip_version} ) {
                 push @results,
```

**Comment by d4nys3k** [comment]
[RFC 2182](https://www.rfc-editor.org/rfc/rfc2182) in paragraph 5 states:
> It is recommended that three servers be provided for most organisation level zones, **with at least one** which must be well removed from the others.

So if you have two of three nameservers in single subnet and last one in different network, it's conforming with RFC 2182.

**Comment by matsduf** [comment]
Thanks for pointing that out. It does not seem that RFC 2182 is fully consistent in its recommendations (it is a BCP). Please write you suggestion for the "algorithm" for the test case. If you have any suggestions for the messages (section "Summary") please include those.

**Comment by d4nys3k** [comment]
I agree, texting, especially in older RFCs, can be confusing sometimes. 

In [CONNECTIVITY04](https://github.com/zonemaster/zonemaster/blob/develop/docs/public/specifications/tests/Connectivity-TP/connectivity04.md) specification, under **Test procedure**, paragraphs 5 and 6 I suggest following change:

> i:    For each IP prefix in the set that has two or more members

replace by: _If all  addresses share same IP prefix_

> ii:    For all IP prefixes in the set that have exactly one member,

replace by: _If there's at least one address in different IP prefix_

Or in some other way in the spirit of RFC 2182 paragraph, I admit that specifying it appropriately is not completely easy even for me. But the goal is probably simple - zonemaster should not report an error where the configuration does not contradict RFC. Above is my proposal to `Connectivity.pm` code change, I think code describes my thoughts in the best way. Of course, change can be more complex, but probably it isn't necessary.

**Comment by LeffeG** [comment]
I tried the patch above and it looks ok so far in my tests.

**Comment by matsduf** [comment]
This issue has also been raised in #1221.

**Comment by rosjat** [comment]
I saw zonemaster released a new version but this issue persist if I run a test.

Will this be addressed at all in the near future?

Cheers

**Comment by matsduf** [comment]
This issue will be addressed, but not in the upcoming release, v2023.2, that will be published in the beginning of January. After that release I will start working on the issue. Thank you for your patience.
